File: golang/executors/sqlx_ex/post_document.go
package sqlx_ex

import (
	"bench-pg-go/executors/sqlx_ex/model"
	"bench-pg-go/model/domain"
	"database/sql"
	"errors"
	"fmt"
	"github.com/jmoiron/sqlx"
	"github.com/shopspring/decimal"
	"time"
)

func postDocument(tx *sqlx.Tx, document *domain.Document) error {

	balance := decimal.Zero
	for _, line := range document.Lines {
		balance = balance.Add(line.Amount)
	}
	if !balance.Equal(decimal.Zero) {
		return errors.New(fmt.Sprintf("balance is %s, must be zero", balance.String()))
	}

	periodYp := domain.PeriodFromDate(document.PostingDate).YearPeriod()
	userId := document.UpdatedBy.Id()
	updatedAt := time.Now()

	items := make([]model.LedgerItemEntity, 0, len(document.Lines))
	for _, line := range document.Lines {
		debt := decimal.Zero
		if line.DebtCredit == domain.DEBT {
			debt = line.Amount
		}
		credit := decimal.Zero
		if line.DebtCredit == domain.CREDIT {
			credit = line.Amount.Neg()
		}
		quantity := decimal.Zero
		if line.Quantity != nil {
			quantity = *line.Quantity
		}
		var materialId sql.Null[string]
		if line.Material != nil {
			materialId.V = *line.Material
			materialId.Valid = true
		}
		var businessPartnerId sql.Null[string]
		if line.BusinessPartner != nil {
			businessPartnerId.V = *line.BusinessPartner
			businessPartnerId.Valid = true
		}
		item := model.LedgerItemEntity{
			Id:                0,
			Period:            periodYp,
			DocNo:             document.DocNo,
			PostingDate:       document.PostingDate.String(),
			AccountId:         line.Account,
			MaterialId:        materialId,
			BusinessPartnerId: businessPartnerId,
			DebtCredit:        line.DebtCredit,
			Amount:            line.Amount,
			Debt:              debt,
			Credit:            credit,
			Quantity:          quantity,
			UpdatedBy:         userId,
			UpdatedAt:         updatedAt,
		}
		items = append(items, item)
	}

	_, err := tx.NamedExec(
		`insert into fin_ledger_items ( period, doc_no, posting_date,
                          account_id, business_partner_id, material_id, debt_credit,
                          amount, debt, credit, quantity,
                          updated_by, updated_at)
           		 values (:period, :doc_no, :posting_date,
                         :account_id, :business_partner_id, :material_id, :debt_credit,
                         :amount, :debt, :credit, :quantity,
                         :updated_by, :updated_at)`,
		items,
	)
	if err != nil {
		return err
	}

	return nil
}

File: rust/src/executors/sqlx/post_document.rs
use crate::model::domain::{DebtCredit, Document, Period};
use chrono::Utc;
use sqlx::types::Decimal;
use sqlx::{Postgres, Transaction};

// pub async fn post_document(
//     tx: &mut Transaction<'_, Postgres>,
//     document: Document,
// ) -> Result<(), String> {
//     let balance = document
//         .lines
//         .iter()
//         .map(|line| line.amount)
//         .sum::<Decimal>();
//     if balance != Decimal::from(0) {
//         return Err(format!("Balance = {} must be zero.", balance));
//     }
//
//     let period_yp = Period::from(document.posting_date).year_period();
//     let user_id = document.updated_by.id();
//     let updated_at = Utc::now();

// let len = document.lines.len();
// let periods = (0..len).map(|_| period_yp).collect::<Vec<_>>();
// let doc_nos = (0..len).map(|_| &document.doc_no).collect::<Vec<_>>();
// let posting_dates = (0..len).map(|_| &document.posting_date).collect::<Vec<_>>();
// let user_ids = (0..len).map(|_| &user_id).collect::<Vec<_>>();
// let updated_ats = (0..len).map(|_| &updated_at).collect::<Vec<_>>();
// let accounts = document.lines.iter().map(|line| line.account).collect::<Vec<_>>();
// let business_partners = document.lines.iter().map(|line| line.business_partner.as_ref().map(|bp| bp.id())).collect::<Vec<_>>();
// let materials = document.lines.iter().map(|line| line.material.as_ref().map(|m| m.id())).collect::<Vec<_>>();
// let debt_credits = document.lines.iter().map(|line| line.debt_credit.to_db_option()).collect::<Vec<_>>();
// let amounts = document.lines.iter().map(|line| line.amount).collect::<Vec<_>>();
// let debts = document.lines.iter().map(|line| {
//     if line.debt_credit == DebtCredit::Debt {line.amount} else {Decimal::from(0)}
// }).collect::<Vec<_>>();
// let credits = document.lines.iter().map(|line| {
//     if line.debt_credit == DebtCredit::Credit {-line.amount} else {Decimal::from(0)}
// }).collect::<Vec<_>>();
// let quantities = document.lines.iter().map(|line| line.quantity.unwrap_or(Decimal::from(0))).collect::<Vec<_>>();
//
// sqlx::query(r#"
//             insert into fin_ledger_items (
//               period, doc_no, posting_date,
//               account_id, business_partner_id, material_id, debt_credit,
//               amount, debt, credit, quantity,
//               updated_by, updated_at
//             ) select * from unnest(
//                 $1::int4[],
//                 $2::varchar[],
//                 $3::date[],
//                 $4::varchar[],
//                 $5::varchar[],
//                 $6::varchar[],
//                 $7::varchar[],
//                 $8::numeric[],
//                 $9::numeric[],
//                 $10::numeric[],
//                 $11::numeric[],
//                 $12::varchar[],
//                 $13::timestamptz[]
//             )"#
// )
//     .bind(periods)
//     .bind(doc_nos)
//     .bind(posting_dates)
//     .bind(accounts)
//     .bind(business_partners)
//     .bind(materials)
//     .bind(debt_credits)
//     .bind(amounts)
//     .bind(debts)
//     .bind(credits)
//     .bind(quantities)
//     .bind(user_ids)
//     .bind(updated_ats
// )
//     .execute(&mut **tx)
//     .await
//     .map_err(|e| e.to_string())?;
// }


pub async fn post_document(
    tx: &mut Transaction<'_, Postgres>,
    document: Document,
) -> Result<(), String> {
    let balance = document
        .lines
        .iter()
        .map(|line| line.amount)
        .sum::<Decimal>();
    if balance != Decimal::from(0) {
        return Err(format!("Balance = {} must be zero.", balance));
    }

    let period_yp = Period::from(document.posting_date).year_period();
    let user_id = document.updated_by.id();
    let updated_at = Utc::now();
    
    for line in document.lines {
        let (debt, credit) = match line.debt_credit {
            DebtCredit::Debt => (&line.amount.clone(), &Decimal::from(0)),
            DebtCredit::Credit => (&Decimal::from(0), &-line.amount.clone()),
        };

        sqlx::query!(
            r#"insert into fin_ledger_items ( period, doc_no, posting_date,
                          account_id, business_partner_id, material_id, debt_credit,
                          amount, debt, credit, quantity,
                          updated_by, updated_at)
           values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13) "#,
            period_yp,
            document.doc_no,
            document.posting_date,
            line.account,
            line.business_partner.map(|bp| bp.id()),
            line.material.map(|m| m.id()),
            line.debt_credit.to_db_option(),
            &line.amount,
            debt,
            credit,
            line.quantity.unwrap_or(Decimal::from(0)),
            &user_id,
            &updated_at
        )
        .execute(&mut **tx)
        .await
        .map_err(|e| e.to_string())?;
    }

    Ok(())
}
